pallets/gear/src/manager/journal.rs

```diff
@@ -415,4 +421,72 @@ where
         QueueOf::<T>::requeue(dispatch)
             .unwrap_or_else(|e| unreachable!("Message queue corrupted! {:?}", e));
     }
+
+    fn reserve_gas(
+        &mut self,
+        message_id: MessageId,
+        reservation_id: ReservationId,
+        program_id: ProgramId,
+        amount: u64,
+        duration: u32,
+    ) {
+        log::debug!(
+            "Reserved: {:?} from {:?} with {:?} for {} blocks",
+            amount,
+            message_id,
+            reservation_id,
+            duration
+        );
+
+        let hold = HoldBound::<T>::by(CostsPerBlockOf::<T>::reservation())
+            .duration(BlockNumberFor::<T>::from(duration));
+
+        // Validating holding duration.
+        if hold.expected_duration().is_zero() {
+            unreachable!("Threshold for reservation invalidated")
+        }
+
+        let total_amount = amount.saturating_add(hold.lock());
+
+        GasHandlerOf::<T>::reserve(message_id, reservation_id, total_amount)
+            .unwrap_or_else(|e| unreachable!("GasTree corrupted: {:?}", e));
+
+        GasHandlerOf::<T>::lock(reservation_id, hold.lock())
+            .unwrap_or_else(|e| unreachable!("GasTree corrupted! {:?}", e));
+
+        TaskPoolOf::<T>::add(
+            hold.deadline(),
+            ScheduledTask::RemoveGasReservation(program_id, reservation_id),
+        )
+        .unwrap_or_else(|e| unreachable!("Scheduling logic invalidated! {:?}", e));
+    }
+
+    fn unreserve_gas(&mut self, reservation_id: ReservationId, program_id: ProgramId, bn: u32) {
+        <Self as TaskHandler<T::AccountId>>::remove_gas_reservation(
+            self,
+            program_id,
+            reservation_id,
+        );
+
+        let _ = TaskPoolOf::<T>::delete(
+            BlockNumberFor::<T>::from(bn),
+            ScheduledTask::RemoveGasReservation(program_id, reservation_id),
+        );
+    }
+
+    fn update_gas_reservation(&mut self, program_id: ProgramId, reserver: GasReserver) {
+        let pid = program_id.into_origin();
+        let prog = common::get_program(pid).unwrap_or_else(|| {
+            unreachable!("gas reservation update guaranteed to be called only on existing program")
+        });
+        if let Program::Active(mut prog) = prog {
+            prog.gas_reservation_map = reserver.into_map(|duration| {
+                HoldBound::<T>::by(CostsPerBlockOf::<T>::reservation())
+                    .duration(BlockNumberFor::<T>::from(duration))
+                    .deadline()
```

Comment:
deadline is for scheduler
default logic should be based on expected() fn

Reply:
otherwise `<Self as TaskHandler<T::AccountId>>::remove_gas_reservation` won't work